data/src/main/java/com/microsoft/azure/kusto/data/BaseClient.java

```diff
@@ -0,0 +1,188 @@
+import com.microsoft.azure.kusto.data.http.HttpRequestBuilder;
+import com.microsoft.azure.kusto.data.http.HttpStatus;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.conn.EofSensorInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.invoke.MethodHandles;
+import java.util.Optional;
+import java.util.zip.DeflaterInputStream;
+import java.util.zip.GZIPInputStream;
+
+public abstract class BaseClient implements Client, StreamingClient {
+
+    private static final int MAX_REDIRECT_COUNT = 1;
+
+    // Make logger available to implementations
+    protected static final Logger LOGGER = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final HttpClient httpClient;
+
+    public BaseClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    protected String post(HttpRequest request) throws DataServiceException {
+
+        // Todo: Add async version of this method
+
+        // Execute and get the response
+        try (HttpResponse response = httpClient.sendSync(request, Context.NONE)) {
+            Optional<HttpHeader> contentEncoding = Optional.ofNullable(response.getHeaders().get(HttpHeaderName.CONTENT_ENCODING));
+            String responseBody = contentEncoding.isPresent() && (contentEncoding.get().getValue().contains("gzip")) ? response.getBodyAsInputStream()
+                    .map(Utils::gzipedInputToString)
+                    .block()
+                    : response.getBodyAsString().block();
+
+            if (responseBody != null) {
+                switch (response.getStatusCode()) {
+                    case HttpStatus.OK:
+                        return responseBody;
+                    case HttpStatus.TOO_MANY_REQS:
+                        throw new ThrottleException(request.getUrl().toString());
+                    default:
+                        throw createExceptionFromResponse(request.getUrl().toString(), response, null, responseBody);
+                }
+            }
+        }
+
+        return null;
+    }
+
+    protected InputStream postToStreamingOutput(HttpRequest request) throws DataServiceException {
+        return postToStreamingOutput(request, 0);
+    }
+
+    protected InputStream postToStreamingOutput(HttpRequest request, int redirectCount) throws DataServiceException {
+
+        boolean returnInputStream = false;
+        String errorFromResponse = null;
+
+        HttpResponse httpResponse = null;
+        try {
+
+            // Todo: Implement async version of this method
+
+            httpResponse = httpClient.sendSync(request, Context.NONE);
+
+            int responseStatusCode = httpResponse.getStatusCode();
+
+            if (responseStatusCode == HttpStatus.OK) {
+                InputStream contentStream = new EofSensorInputStream(new CloseParentResourcesStream(httpResponse), null);
```

Comment:
Do we still need these two stream wrappers, even with the current http client?

Reply:
@AsafMah I am not sure. I kept them in place because I couldn't answer that myself. Would tests cover this if I changed it?

Reply:
@AsafMah removed in [this commit](https://github.com/Azure/azure-kusto-java/pull/335/commits/54d1ec5d6f07ae58cc30711169f016f3de2a2fa1)

Reply:
@ohadbitt this was in postToStreamingOutput, so we'll probably want to double triple check that streaming queries are okay. I verified that all tests I can run still pass.